### DISCUSSION TITLE
Failed to look up dependency defaults:integer

### DISCUSSION THREAD

**Discussion by mirland** [renovatebot/renovate#12240] [Request Help]
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Please select which platform you are using if self-hosting.

github.com

### Describe the bug

Renovate stopped to work some time ago and I've not changed anything, so I think some change in your side broke something.

This is the error that the logs it's showing: `Failed to look up dependency defaults:integer`.

 The project that is giving the error is similar to https://github.com/xmartlabs/gong/tree/master 
 It's not exactly the same project but the dependencies are defined in the same way.
Keep in mind that you should check master, which is an old version of main.

The versions are managed in a file similar to https://github.com/xmartlabs/gong/blob/master/versions.gradle.

### Relevant debug logs

<details><summary>
"deps": [
          {
            "depName": "defaults:integer",
            "currentValue": "integer",
            "managerData": {
              "fileReplacePosition": 9722,
              "packageFile": "app/build.gradle"
            },
            "fileReplacePosition": 9722,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "https://dl.google.com/android/maven2/",
              "https://plugins.gradle.org/m2/",
              "https://kotlin.bintray.com/kotlinx/",
              "https://zendesk.jfrog.io/zendesk/repo",
              "https://s3.amazonaws.com/repo.commonsware.com",
              "https://jitpack.io"
            ],
            "depIndex": 0,
            "warnings": [
              {
                "topic": "defaults:integer",
                "message": "Failed to look up dependency defaults:integer"
              }
            ],
            "versioning": "gradle",
            "updates": []
          }
        ]
      },

</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by mirland** [comment]
A minor comment is that the app module depends on a git submodule. Maybe the submodule it's not cloned in the right way and it's related to that. I'm not sure, but it's something to keep in mind

**Reply by viceice** [reply]
`clonesubmodules` is now false by default, so you explicitly need to enable it in your repo config.

https://docs.renovatebot.com/configuration-options/#clonesubmodules
https://github.com/renovatebot/renovate/releases/tag/26.0.0

**Comment by viceice** [comment]
`clonesubmodules` is now false by default, so you explicitly need to enable it in your repo config.

https://docs.renovatebot.com/configuration-options/#clonesubmodules
https://github.com/renovatebot/renovate/releases/tag/26.0.0

**Comment by mirland** [comment]
@viceice Thanks for your response, I've enabled the submodule property. However, it's still not working. I'm having the same issue:

```
{
                "topic": "defaults:integer",
                "message": "Failed to look up dependency defaults:integer"
              }
```  

How can I know exactly what's it's happening?

**Reply by viceice** [reply]
I think you need to create a minimal reproduction repo, so somebody can run a debug session and check what happened.

Are there any other related log messages to that dependency?

Which manager is extracting this dep?

**Comment by mirland** [comment]
I checked again and I have only that warning:

<details><summary>Logs</summary>

```
DEBUG: Looking up defaults:integer in repository https://repo.maven.apache.org/maven2/
DEBUG: Content is not found for Maven url: https://repo.maven.apache.org/maven2/defaults/integer/maven-metadata.xml
DEBUG: Looking up defaults:integer in repository https://dl.google.com/android/maven2/
DEBUG: Content is not found for Maven url: https://dl.google.com/android/maven2/defaults/integer/maven-metadata.xml
DEBUG: Looking up defaults:integer in repository https://plugins.gradle.org/m2/
DEBUG: Content is not found for Maven url: https://plugins.gradle.org/m2/defaults/integer/maven-metadata.xml
DEBUG: Looking up defaults:integer in repository https://zendesk.jfrog.io/zendesk/repo/
DEBUG: Content is not found for Maven url: https://zendesk.jfrog.io/zendesk/repo/defaults/integer/maven-metadata.xml
DEBUG: Looking up defaults:integer in repository https://s3.amazonaws.com/repo.commonsware.com/
DEBUG: Content is not found for Maven url: https://s3.amazonaws.com/repo.commonsware.com/defaults/integer/maven-metadata.xml
DEBUG: Looking up defaults:integer in repository https://jitpack.io/
DEBUG: Content is not found for Maven url: https://jitpack.io/defaults/integer/maven-metadata.xml
DEBUG: Failed to look up dependency defaults:integer (defaults:integer)(packageFile="app/build.gradle", dependency="defaults:integer")
DEBUG: Package releases lookups complete
{
  "baseBranch": "develop"
}
DEBUG: branchifyUpgrades
DEBUG: 0 flattened updates found:
DEBUG: Returning 0 branch(es)
DEBUG: config.repoIsOnboarded=true
DEBUG: packageFiles with updates
{
  "config": {
    "github-actions": [
      {
        "packageFile": ".github/workflows/compile.yml",
        "deps": [
          {
            "depName": "actions/checkout",
            "commitMessageTopic": "{{{depName}}} action",
            "datasource": "github-tags",
            "versioning": "docker",
            "depType": "action",
            "replaceString": "actions/checkout@v2",
            "autoReplaceStringTemplate": "{{depName}}@{{#if newDigest}}{{newDigest}}{{#if newValue}} # renovate: tag={{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}",
            "currentValue": "v2",
            "depIndex": 0,
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/actions/checkout",
            "currentVersion": "v2",
            "fixedVersion": "v2"
          },
          {
            "depName": "actions/cache",
            "commitMessageTopic": "{{{depName}}} action",
            "datasource": "github-tags",
            "versioning": "docker",
            "depType": "action",
            "replaceString": "actions/cache@v2",
            "autoReplaceStringTemplate": "{{depName}}@{{#if newDigest}}{{newDigest}}{{#if newValue}} # renovate: tag={{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}",
            "currentValue": "v2",
            "depIndex": 1,
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/actions/cache",
            "currentVersion": "v2",
            "fixedVersion": "v2"
          },
          {
            "depName": "actions/upload-artifact",
            "commitMessageTopic": "{{{depName}}} action",
            "datasource": "github-tags",
            "versioning": "docker",
            "depType": "action",
            "replaceString": "actions/upload-artifact@v2",
            "autoReplaceStringTemplate": "{{depName}}@{{#if newDigest}}{{newDigest}}{{#if newValue}} # renovate: tag={{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}",
            "currentValue": "v2",
            "depIndex": 2,
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/actions/upload-artifact",
            "currentVersion": "v2",
            "fixedVersion": "v2"
          },
          {
            "depName": "actions/upload-artifact",
            "commitMessageTopic": "{{{depName}}} action",
            "datasource": "github-tags",
            "versioning": "docker",
            "depType": "action",
            "replaceString": "actions/upload-artifact@v2",
            "autoReplaceStringTemplate": "{{depName}}@{{#if newDigest}}{{newDigest}}{{#if newValue}} # renovate: tag={{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}",
            "currentValue": "v2",
            "depIndex": 3,
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/actions/upload-artifact",
            "currentVersion": "v2",
            "fixedVersion": "v2"
          }
        ]
      },
      {
        "packageFile": ".github/workflows/submodule_commit.yml",
        "deps": [
          {
            "depName": "actions/checkout",
            "commitMessageTopic": "{{{depName}}} action",
            "datasource": "github-tags",
            "versioning": "docker",
            "depType": "action",
            "replaceString": "actions/checkout@v2",
            "autoReplaceStringTemplate": "{{depName}}@{{#if newDigest}}{{newDigest}}{{#if newValue}} # renovate: tag={{newValue}}{{/if}}{{/if}}{{#unless newDigest}}{{newValue}}{{/unless}}",
            "currentValue": "v2",
            "depIndex": 0,
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/actions/checkout",
            "currentVersion": "v2",
            "fixedVersion": "v2"
          }
        ]
      }
    ],
    "gradle": [
      {
        "packageFile": "gradle.properties",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "build.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "settings.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "versions.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "app/build.gradle",
        "datasource": "maven",
        "deps": [
          {
            "depName": "defaults:integer",
            "currentValue": "integer",
            "managerData": {
              "fileReplacePosition": 9722,
              "packageFile": "app/build.gradle"
            },
            "fileReplacePosition": 9722,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "https://dl.google.com/android/maven2/",
              "https://plugins.gradle.org/m2/",
              "https://zendesk.jfrog.io/zendesk/repo",
              "https://s3.amazonaws.com/repo.commonsware.com",
              "https://jitpack.io"
            ],
            "depIndex": 0,
            "updates": [],
            "warnings": [
              {
                "topic": "defaults:integer",
                "message": "Failed to look up dependency defaults:integer"
              }
            ],
            "versioning": "gradle"
          }
        ]
      },
      {
        "packageFile": "core/build.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "crop-image/build.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "font-picker/build.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "instagram/build.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "scripts/dependency_updates.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "scripts/read_properties.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "scripts/snapshot_publish.gradle",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "scripts/stop_ship_check_xmls.gradle.kts",
        "datasource": "maven",
        "deps": []
      },
      {
        "packageFile": "scripts/versioning.gradle",
        "datasource": "maven",
        "deps": []
      }
    ],
    "gradle-wrapper": [
      {
        "packageFile": "gradle/wrapper/gradle-wrapper.properties",
        "deps": [
          {
            "depName": "gradle",
            "currentValue": "7.2",
            "replaceString": "https\\://services.gradle.org/distributions/gradle-7.2-all.zip",
            "datasource": "gradle-version",
            "versioning": "gradle",
            "depIndex": 0,
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/gradle/gradle",
            "homepage": "https://gradle.org",
            "currentVersion": "7.2",
            "fixedVersion": "7.2"
          }
        ]
      }
    ]
  }
}
```
</details>

I'll run it in Gong's repo and I will try to reproduce it. However, shouldn't I see something in the log? How can I see more info?

**Reply by viceice** [reply]
Where should it find `defaults:integer`? It seems you should simply disable that dep, as i don't see anything renovate can update. 🤷‍♂️

**Reply by mirland** [reply]
I cannot find `defaults:integer` in my project. (I searched in the whole project, I don't know what `defaults:integer` means).
After that, I searched `integer` in my build gradle files (*.gradle) and I found something interesting.

```
apply plugin: "com.github.konifar.gradle.unused-resources-remover"

unusedResourcesRemover {
    extraRemovers = [
            createXmlValueRemover("defaults", "integer", "integer")
    ]
    excludeNames = ["attrs_proportional_layout.xml"]
}
```

I'm using [this plugin](https://github.com/konifar/gradle-unused-resources-remover-plugin), I don't know why, but it seems that renovate is having issues with that plugin (It was working as expected some time ago, I mean, I added that plugin and I didn't update for a long time).

**Reply by rarkins** [reply]
Are you able to reproduce it in a minimal public repo now?

**Reply by mirland** [reply]
Yes, the issue is related to what I mentioned. I forked [gong](https://github.com/mirland/gong) and I integrated renovate. I waited and it ran ok. It created 2 prs (https://github.com/mirland/gong/pull/1 and https://github.com/mirland/gong/pull/2).

Then, I integrated "Unused Resources Remover for Android" plugin, you can check [this commit](https://github.com/mirland/gong/commit/5801f6b45c48a457dd6e4c9a882c182ee54c17e6) and renovate starting to give exactly the same issue. Currently it's not working anymore!

So, I can confirm it's an issue on renovate side because the plugin should not be considered as a dependency.

BTW: I tried to ignore the dependency `"defaults:integer"` in the configuration, but renovate it's still not working.

**Reply by rarkins** [reply]
I was referring more specifically to this form of reproduction: https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md

**Reply by mirland** [reply]
can you check [this repo](https://github.com/mirland/renovate-plugin-issue-mrr ) now, please?

**Reply by rarkins** [reply]
@zharinov can you take a look and confirm if this can be fixed?

**Reply by zharinov** [reply]
Our parser wrongly treats string `createXmlValueRemover("defaults", "integer", "integer")` as the dependency. Though we need to fix parser behavior, I'm not sure that's the root cause.

I wasn't able to reproduce the problem with my fork of reproduction repo, but maybe can you try remove this string from `app/build.gradle` file and see if it makes any difference?

**Reply by mirland** [reply]
If you remove the plugin, renovate doesn't work. 
commits: 
- https://github.com/mirland/renovate-plugin-issue-mrr/commit/7a4a24ed896c3714cf6334f3436dfb01ef5ebbf6
- https://github.com/mirland/renovate-plugin-issue-mrr/commit/1a65219fd3454b6920aab48edabdfd7682c17a7c
- https://github.com/mirland/renovate-plugin-issue-mrr/commit/f1a3cc7291fce4414f5d27ddbc80490b651d2774


<details><summary>Logs</summary>
```
{
        "packageFile": "app/build.gradle",
        "datasource": "maven",
        "deps": [
          {
            "depName": "junit:junit",
            "currentValue": "4.+",
            "managerData": {
              "fileReplacePosition": 1059,
              "packageFile": "app/build.gradle"
            },
            "fileReplacePosition": 1059,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "https://dl.google.com/android/maven2/",
              "https://plugins.gradle.org/m2/",
              "https://jcenter.bintray.com/"
            ],
            "depIndex": 0,
            "warnings": [],
            "versioning": "gradle",
            "sourceUrl": "https://github.com/junit-team/junit4",
            "homepage": "http://junit.org",
            "currentVersion": "4.13.2",
            "updates": []
          }
        ]
      }
    ],
    ```
</details>

So I think there're two different problems
1- The parser reats string createXmlValueRemover("defaults", "integer", "integer") as the dependency. 
2- The include the versions in a common [version file](https://github.com/mirland/renovate-plugin-issue-mrr/blob/main/versions.gradle) is not working (It was working some time ago, the last dependency that I merge was from Aug). I confirmed it "forking" again this repo, and using a new setup without the remover plugin https://github.com/mirland/renovate-plugin-issue-mrr-2 and it's not working

**Reply by mirland** [reply]
@ZauberNerd Do you have some updates? can you confirm if this will be fixed or not? Renovates stopped work in August, and I have to know if I need to find an alternative or not

**Reply by rarkins** [reply]
@mirland once you have a confirmed minimal reproduction - for example a public github.com repository which reproduces the error every time with the app - then it's ok to have a second attempt at a bug report issue. But please be realistic that as long as this only affects you/your company, and you are not paying anything, or able to contribute a fix yourselves, then it's likely not going to be prioritized.

**Reply by mirland** [reply]
@rarkins Hi thanks for the clarification! I think that was not clear at the beginning. 
Some comments:
1- I've sent the minimal reproduction repository, is it ok right? 

> but please be realistic that as long as this only affects you/your company

2- This way of declaring dependencies [is used by google](https://github.com/android/architecture-components-samples/blob/main/versions.gradle) (21k stars in that repo) , it's not a standard (because it didn't exist a standard) but a lot of people use this way. So I'm sure it's not affecting only my company. The main issue is that you have to spend a lot of time researching what's happening, creating a MRR, and then reporting the issue. That's why I think I'm the first person that reported it. I could spend that time migrating to another tool, but I preferred spending that time improving this tool.

>  able to contribute a fix yourselves,

3- In my opinion, reporting the issue is a way of contributing. Then, it can be hard to fix the issue, if it's the first issue that you solve in the repo, and you have no support.

Again, thanks for the clarification, and have a great day

**Reply by zharinov** [reply]
@mirland I'm creating some new parsing tool that can be used for handling dependencies in such elaborate build systems as Gradle is (from my experience, it's is the most complex one). However, I believe there is some room for quick improvement for our current implementation. Feel free to subscribe to the issue I've just created to address your concerns: #12518

**Reply by zharinov** [reply]
Today we've released improvements for gradle support

**Reply by mirland** [reply]
It worked perfectly. Thanks for the hard work :muscle:
